tests/paths.py

```diff
@@ -0,0 +1,94 @@
+# Copyright (c) 2023 The MobileCoin Foundation
+
+from filecmp import dircmp
+from pathlib import Path
+import seedir as sd
+
+#: The root of this repository. This can be used as the `template` parameter to
+#: `cookiecutter` specifying the directory to the desired template as one would
+#: do on the command line.
+TEMPLATE_ROOT = Path(__file__).parent.parent
+
+
+class ReprPath:
+    """
+    A `Path` like object which provides the directory contents or file contents
+    in the `__repr__` method.
+
+    Directories will use a unix `tree` like output. Files will be the raw file
+    contents.
+    """
+
+    def __init__(self, path):
+        self.path = Path(path)
+
+    def __repr__(self):
+        if self.path.is_dir():
+            tree = sd.seedir(self.path, printout=False)
+            return tree
+
+        with open(self.path) as f:
+            return f.read()
+
+    def __truediv__(self, other):
+        return ReprPath(self.path / other)
+
+
+class DiffDir:
+    """
+    Provides a way of comparing directories for use with pytest.
+
+    Implements an `==` operator that looks at the directory contents of
+    both `a` and `b` in `a == b`.
+
+    Uses a stateful memory of the last `==` to update the `__repr__` of the `a`
+    and `b` used in the comparison. The `__repr__`s contain only the directories
+    and files that are different between `a` and `b`.
+    """
+
+    def __init__(self, path):
+        self.path = path
+        self.repr = str(self.path)
+
+    def __eq__(self, other):
+        _dirs = dircmp(self.path, other.path)
+        left, right = diff_dirs(_dirs)
+
+        if left or right:
+            self.repr = "\n".join(str(l) for l in left)
+            other.repr = "\n".join(str(r) for r in right)
+            return False
```

Comment:
I'm not fond of this behavior of updating state after a `==` operation.
There is https://docs.pytest.org/en/7.1.x/how-to/assert.html#defining-your-own-explanation-for-failed-assertions. but overriding all `Path` comparisons seemed like it might be a bit confusing.
It could also be that I'm too attached to pytests `assert a == b` default and this should be `some_assertion_method(left, right)`

Reply:
Changed to use the custom explanation by keying off the `DiffPath` class type